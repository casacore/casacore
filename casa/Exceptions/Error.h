//# Error.h: Base class for all Casacore errors
//# Copyright (C) 1993,1994,1995,1999,2000,2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef CASA_ERROR_H
#define CASA_ERROR_H



#include <sys/types.h>
#include <casa/aips.h>
#include <casa/BasicSL/String.h>
#include <casa/OS/Mutex.h>
#include <exception>


namespace casa { //# NAMESPACE CASA - BEGIN

// Throw the given exception with a string composed of various arguments.
// E.g.
// <srcblock>
//    CASATHROW (AipsError, "integer=" << myint << ", float=" << myfloat);
// </srcblock>
#define CASATHROW(exc, arg) do {     \
    std::ostringstream casa_log_oss; \
    casa_log_oss << arg;             \
    throw exc(casa_log_oss.str());   \
  } while (0)

// Throw an AipsError exception containing the file and line.
#define ThrowCc(m) throw AipsError ((m), __FILE__, __LINE__)

// Throw an AipsError exception if the condition is true.
#define ThrowIf(c,m) {if (c) {casa::AipsError::throwIf (True, (m), __FILE__, __LINE__, __PRETTY_FUNCTION__);}}

// Throw an AipsError exception if the system error code is not 0.
// It adds the message for that error code to the exception text.
#define ThrowIfError(c,m) {if (c) {casa::AipsError::throwIfError (True, (m), __FILE__, __LINE__, __PRETTY_FUNCTION__);}}

// Repackage and rethrow an AipsError exception.
#define Rethrow(e,m) {throw casa::AipsError::repackageAipsError ((e),(m),__FILE__,__LINE__, __PRETTY_FUNCTION__);}


// <summary>Base class for all Casacore library errors</summary>
// <use visibility=export>
//
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>
//
// <prerequisite>
//   <li> ExcpError
// </prerequisite>
//
// <synopsis>
//  This is the base class for all of the Casacore error classes. Because
//  all of the errors have a common base class, any error can be caught
//  with a single catch statement.
//
//  This class has a string which allows error messages to be propagated.
//
//  <note role=tip> The string member must be handled very carefully because
//        string is also derived from cleanup, thus the 
//        <src>message.makePermanent()</src> call in the implementation of
//        the constructors. This prevents the String from being cleaned up
//        in the middle of an exception.
//  </note>
//
// </synopsis>
//
// <example>
// <srcblock>
//      throw(AipsError("SOME STRING"));
// </srcblock>
// </example>
//
// <todo asof="">
// </todo>

class AipsError: public std::exception
{
public:

  enum Category {
    BOUNDARY, INITIALIZATION, INVALID_ARGUMENT, CONFORMANCE,
    ENVIRONMENT, SYSTEM, PERMISSION, GENERAL
  };

  //
  // Simply returns the stored error message.
  //
  virtual const char* what() const throw()
    { return(message.c_str()); }
  const String &getMesg() const
    { return(message); }
  String getStackTrace() const;
  AipsError::Category getCategory( ) const
    { return(category); }

  // Append a message. This is used by LogIO when an exception is logged.
  // The message is const to be able to use it for a temporary exception.
  void setMessage (const String& msg) const
    { const_cast<AipsError*>(this)->message = msg; }

  // Creates an AipsError and initializes the error message from
  // the parameter.
  // <group>
  AipsError (const Char *str, Category c = GENERAL);
  AipsError (const String &str, Category c = GENERAL);
  AipsError (const String &msg, const String &filename, uInt lineNumber,
             Category c = GENERAL);
  AipsError (Category c = GENERAL);
  // </group>

  //
  // Destructor which does nothing.
  //
  ~AipsError() throw();

  // Get or clear the stacktrace info.
  // <group>
  static void getLastInfo (String & message, String & stackTrace);
  static String getLastMessage ();
  static String getLastStackTrace ();
  static void clearLastInfo ();
  // </group>

  // Repackage an exception.
  static AipsError repackageAipsError (AipsError& error, 
                                       const String& message,
                                       const char* file,
                                       Int line,
                                       const char* func);

  // Throw if the condition is true.
  static void throwIf (Bool condition, const String& message,
                       const char* file, Int line,
                       const char* func = "");

  // Throw if the system error code is not 0.
  static void throwIfError (Int errorCode, const String& prefix,
                            const char* file, Int line,
                            const char* func = "");

protected:
  // Add the stack trace to the message (if USE_STACKTRACE is set).
  void addStackTrace ();

  String message;
  Category category;
  String stackTrace;
};


// <summary>Allocation errors</summary>
// <use visibility=export>
//
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>
//
// <synopsis>
//
// This class is used for allocation errors. It adds an extra
// data item, the failed allocation size. Otherwise much the
// same as <src>AipsError</src>.
//
// </synopsis>
//
// <example>
// <srcblock>
//     throw(AllocError("ANY STRING",1024));
// </srcblock>
// </example>
//
// <todo asof="">
// </todo>

class AllocError : public AipsError {
protected:
  size_t Size;
public:
  //
  // This constructor takes the error message and the failed
  // allocation size.
  //
  // <group>
  AllocError(const Char *str, uInt sze) : AipsError(str,SYSTEM), Size(sze) {}
  AllocError(const String &str, uInt sze) : AipsError(str,SYSTEM), Size(sze)  {}
  // </group>

  //
  // This function returns the failed allocation size.
  //
  size_t size() const {return(Size);}

  //
  // Destructor which does nothing.
  //
  ~AllocError() throw();

};


// <summary>Base class for all indexing errors</summary>
// <use visibility=export>
//
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>
//
// <synopsis>
// This class is the base class of all <src>IndexError</src>s. It is
// defined to allow the user to catch any of the many kinds of IndexErrors
// which may be thrown. It can also be thrown itself if returning
// the illegal index value is unimportant.
// </synopsis>
//
// <example>
// <srcblock>
//     throw(IndexError("ANY STRING"));
// </srcblock>
// </example>
//
// <todo asof="">
// </todo>

class IndexError : public AipsError {
public:
  //
  // Creates an GeneralIndexError and initializes the error message from
  // the parameter
  // <group>
  IndexError(const Char *str,Category c=BOUNDARY) : AipsError(str,c) {}
  IndexError(const String &str,Category c=BOUNDARY) : AipsError(str,c) {}
  IndexError(Category c=BOUNDARY) : AipsError(c) {}
  // </group>

  //
  // Destructor which does nothing.
  //
  ~IndexError() throw();
};


// <summary>Index errors returning the bad index</summary>
// <use visibility=export>
//
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>
//
// <synopsis>
// This class is templated to allow generalalized indexes to be returned
// with the error message i.e. the class is templated on the index type.
//
// </synopsis>
//
// <example>
// <srcblock>
//     throw(indexError<int>(3,"ANY STRING"));/
// </srcblock>
// </example>
//
// <todo asof="">
// </todo>

template<class t> class indexError : public IndexError {
protected:
  t oIndex;                 // Offending Index
public:
  //
  // This constructor takes the error message and the index
  // which cause the error to occur.
  //
  // <group>
  indexError(t oI, const Char *str, Category c=BOUNDARY);
  indexError(t oI, const String &str, Category c=BOUNDARY);
  indexError(t oI, Category c=BOUNDARY) : IndexError(c), oIndex(oI) {};
  // </group>

  //
  // Destructor which does nothing.
  //
  ~indexError() throw();
};


// <summary>Duplicate key errors</summary>
// <use visibility=export>
//
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>
//
// <synopsis>
// This class is the base class of all duplicate key errors. It is
// defined to allow the user to catch any of the many kinds of DuplErrors
// which may be thrown. It can also be thrown itself if returning
// the illegal key is unimportant.
// </synopsis>
//
// <example>
// <srcblock>
//    throw(DuplError("ANY STRING"));
// </srcblock>
// </example>
//
// <todo asof="">
// </todo>

class DuplError : public AipsError {
public:
  //
  // Creates an DuplError and initializes the error message from
  // the parameter
  // <group>
  DuplError(Category c=BOUNDARY) : AipsError(c) {}
  DuplError(const Char *str,Category c=BOUNDARY) : AipsError(str,c) {}
  DuplError(const String &str,Category c=BOUNDARY) : AipsError(str,c) {}
  // </group>

  //
  // Destructor which does nothing.
  //
  ~DuplError() throw();
};


// <summary>Duplicate key errors where the bad key is returned</summary>
// <use visibility=export>
//
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>
//
// <synopsis>
//  This template is for generalized duplicate key errors where the template
//  type parameter is the type of the key which caused the error. Because this
//  class is derived from <linkto class=DuplError><src>DuplError</src>
//  </linkto>, the user to catch all duplicate key errors with one catch
//  statement. 
//
// </synopsis>
//
// <example>
//     throw(duplError<int>(4,"ANY STRING"));
// </example>
//
// <todo asof="">
// </todo>

template<class t> class duplError : public DuplError {
protected:
  t oKey;                   // Offending Key
public:
  //
  // This constructs a "duplError" for the offending key, and an
  // optional character string.
  //
  // <group>
  duplError(t oI, const Char *str,Category c=BOUNDARY);
  duplError(t oI, const String &str,Category c=BOUNDARY);
  duplError(t oI,Category c=BOUNDARY) : DuplError(c), oKey(oI) {};
  // </group>

  //
  // Destructor which does nothing.
  //
  ~duplError() throw();
};


// <summary>Exception for an error in a system call</summary>
// <use visibility=export>
//
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>
//
// <synopsis>
// This error is to be used for if a system call returns an error.
// It uses strerror to get the system error message.
// </synopsis>

class SystemCallError : public AipsError
{
public:
  // This constructs a "SystemCallError" from the system call function name
  // and the errno.
  SystemCallError(const String &funcName, int error, Category c=GENERAL);

  SystemCallError (int error, const String &msg, const String &filename,
                   uInt lineNumber, Category c=GENERAL);

  // Destructor which does nothing.
  ~SystemCallError() throw();

  // Get the errno.
  int error() const
    { return itsError; }

  // Get the message belonging to an error.
  static String errorMessage(int error);

private:
  int itsError;
};


// <summary>Exception which halts execution</summary>
// <use visibility=export>
//
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>
//
// <synopsis>
// This error causes an execution to halt regardless. It
// causes execution to halt before the exception can be caught.
// </synopsis>
//
// <example>
// <srcblock>
//     throw(AbortError("ANY STRING"));
// </srcblock>
// </example>
//
// <todo asof="">
// </todo>

class AbortError : public AipsError {
public:
  //
  // This constructs a "AbortError" from the error message.
  //
  // <group>
  AbortError(const Char *str,Category c=GENERAL);
  AbortError(const String &str,Category c=GENERAL);
  // </group>

  //
  // Destructor which does nothing.
  //
  ~AbortError() throw();
};



} //# NAMESPACE CASA - END

#ifdef AIPS_NEEDS_RETHROW
#ifndef CASACORE_NEEDS_RETHROW
#define CASACORE_NEEDS_RETHROW
#endif
#endif

#ifdef CASACORE_NEEDS_RETHROW
#define RETHROW(X) throw(X);
#else
#define RETHROW(X)
#endif

#ifndef CASACORE_NO_AUTO_TEMPLATES
#include <casa/Exceptions/Error.tcc>
#endif //# CASACORE_NO_AUTO_TEMPLATES
#endif
