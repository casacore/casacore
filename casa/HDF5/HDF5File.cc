//# HDF5File.h: An class representing an HDF5 file
//# Copyright (C) 2008
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes
#include <casacore/casa/HDF5/HDF5File.h>
#include <casacore/casa/HDF5/HDF5HidMeta.h>
#include <casacore/casa/HDF5/HDF5Error.h>
#include <casacore/casa/OS/RegularFile.h>

namespace casa { //# NAMESPACE CASA - BEGIN

#ifdef HAVE_HDF5

  HDF5File::HDF5File (const String& name,
		      ByteIO::OpenOption option)
    : itsOption (option),
      itsDelete (False)
  {
    // Disable automatic printing of errors.
    H5Eset_auto2(H5E_DEFAULT, NULL, NULL);
    // Use absolute expanded path name.
    setName (Path(name).absoluteName());
    doOpen();
  }

  HDF5File::~HDF5File()
  {
    close();
    if (itsDelete) {
      RegularFile file(getName());
      file.remove();
    }
  }

  Bool HDF5File::isHDF5 (const String& name)
  {
    // Disable automatic printing of errors.
    H5Eset_auto2(H5E_DEFAULT, NULL, NULL);
    return H5Fis_hdf5 (name.chars());
  }

  void HDF5File::reopenRW()
  {
    if (itsOption != ByteIO::Update) {
      close();
      itsOption = ByteIO::Update;
    }
    reopen();
  }

  void HDF5File::close()
  {
    if (isValid()) {
      H5Fclose (getHid());
      clearHid();
    }
  }
    
  void HDF5File::reopen()
  {
    if (isClosed()) {
      doOpen();
    }
  }

  void HDF5File::flush()
  {
    if (isValid()) {
      if (H5Fflush (getHid(), H5F_SCOPE_LOCAL) < 0) {
	throw HDF5Error ("Failure when flushing file " + getName());
      }
    }
  }

  void HDF5File::doOpen()
  {
    // Use 8 byte offets and blocks of 32768 bytes.
    HDF5HidProperty create_plist (H5Pcreate(H5P_FILE_CREATE));
    H5Pset_sizes(create_plist, 8, 8);
    // Use unbuffered IO.
    HDF5HidProperty access_plist (H5Pcreate(H5P_FILE_ACCESS));
    H5Pset_fapl_sec2(access_plist);
    // Set a cache size of 16 MB.
    // First get old values and use them for the other parameters.
    int mdn;
    size_t ccn, ccb;
    double policy;
    H5Pget_cache(access_plist, &mdn, &ccn, &ccb, &policy);
    ccb = 16*1024*1024;
    H5Pset_cache(access_plist, mdn, ccn, ccb, policy);
    // Open or create the file.
    switch (itsOption) {
    case ByteIO::Old:
      setHid (H5Fopen (getName().chars(), H5F_ACC_RDONLY, access_plist));
      break;
    case ByteIO::Delete:
      itsDelete = True;
      // fall through
    case ByteIO::Update:
    case ByteIO::Append:
      setHid (H5Fopen (getName().chars(), H5F_ACC_RDWR, access_plist));
      itsOption = ByteIO::Update;
      break;
    case ByteIO::Scratch:
      itsDelete = True;
      // fall through
    case ByteIO::New:
      setHid (H5Fcreate (getName().chars(), H5F_ACC_TRUNC,
			 create_plist, access_plist));
      itsOption = ByteIO::Update;
      break;
    case ByteIO::NewNoReplace:
      setHid (H5Fcreate (getName().chars(), H5F_ACC_EXCL,
			 create_plist, access_plist));
      itsOption = ByteIO::Update;
      break;
    }
    if (!isValid()) {
      throw HDF5Error ("Could not open or create HDF5 file " + getName());
    }
  }

#else

  HDF5File::HDF5File (const String& name,
		      ByteIO::OpenOption option)
    : itsOption (option),
      itsDelete (False)
  {
    setName (name);
    doOpen();
  }

  HDF5File::~HDF5File()
  {}

  Bool HDF5File::isHDF5 (const String&)
  {
    return False;
  }

  void HDF5File::reopenRW()
  {}

  void HDF5File::close()
  {}
    
  void HDF5File::reopen()
  {}

  void HDF5File::flush()
  {}

  void HDF5File::doOpen()
  {
    throw HDF5Error("HDF5 support is not compiled into this casacore version");
  }

#endif

}
