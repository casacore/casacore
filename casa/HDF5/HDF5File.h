//# HDF5File.h: An class representing an HDF5 file
//# Copyright (C) 2008
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef CASA_HDF5FILE_H
#define CASA_HDF5FILE_H

//# Includes
#include <casa/HDF5/HDF5Object.h>
#include <casa/BasicSL/String.h>
#include <casa/IO/ByteIO.h>

namespace casa { //# NAMESPACE CASA - BEGIN

  // <summary>
  // A class representing an HDF5 file.
  // </summary>

  // <use visibility=export>

  // <reviewed reviewer="" date="" tests="tHDF5File.cc">
  // </reviewed>

  // <prerequisite>
  //   <li> <a href="http://hdf.ncsa.uiuc.edu">HDF5 system</a>
  // </prerequisite>

  // <synopsis>
  // This class wraps the HDF5 functions to open, create, or close an
  // HDF5 file.
  // If the file is opened as readonly, it is possible to reopen it for
  // read/write (provided the user has the correct privileges).
  // It is also possible to temporarily close the file and reopen it later.
  // <note> It is ensured that the class and the static function <tt>isHDF5</tt>
  // are also defined if HDF5 is not compiled in. </note>
  // </synopsis> 

  // <motivation>
  // It was overkill to use the HDF5 C++ interface. Instead little wrappers
  // have been written. HDF5File can be embedded in a shared pointer making
  // it possible to share an HDF5 file amongst various HDF5Array objects.
  // </motivation>

  class HDF5File : public HDF5Object
  {
  public:
    // Create an HDF5 file object with the given file name (possible tilde
    // or environment variables in it will be expanded).
    // The ByteIO option determines if the file will be created,
    // opened for input and/or output, or possibly deleted by the destructor.
    explicit HDF5File (const String& name,
		       ByteIO::OpenOption = ByteIO::Old);

    // The destructor closes the file and deletes it when it was opened
    // using ByteIO::Scratch or ByteIO::Delete.
    ~HDF5File();

    // Test if the file with the given name is an HDF5 file.
    static Bool isHDF5 (const String& name);

    // Reopen the underlying file for read/write access.
    // Nothing will be done if the stream is writable already.
    // Otherwise it will be reopened and an exception will be thrown
    // if it is not possible to reopen it for read/write access.
    void reopenRW();

    // Is the file writable?
    Bool isWritable() const
      { return itsOption == ByteIO::Update; }

    // Is the file opened for delete?
    Bool isOpenedForDelete() const
      { return itsDelete; }

    // Is the file temporarily closed?
    Bool isClosed() const
      { return getHid()<0; }

    // Close the file (temporarily).
    // Note it will not delete the file; that is only done by the destructor.
    virtual void close();

    // Reopen the file if closed (which may change the HID).
    void reopen();

    // Flush the data to disk.
    void flush();

    // Get or set the chunk cache size (in bytes).
    // Note that all data sets in a file share the cache.
    // <group>
    size_t getChunkCacheSize() const;
    void setChunkCacheSize (size_t nbytes);
    // </group>

  private:
    // Open or create the file.
    void doOpen();


    //# Data members
    ByteIO::OpenOption itsOption;
    String             itsName;
    Bool               itsDelete;

  private:
    // Copy constructor cannot be used.
    HDF5File (const HDF5File& that);

    // Assignment cannot be used.
    HDF5File& operator= (const HDF5File& that);
  };

}

#endif
