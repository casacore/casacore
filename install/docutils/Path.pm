#
#   This is part of cxx2html Version 1.2 Patchlevel 4
#   Created by Darrell Schiebel (drs@nrao.edu)
#
#   This utility, cxx2html, is part of AIPS++, a software project
#   centered at the National Radio Astronomy Observatory.
#
#   Copyright (C) 1995
#   Associated Universities, Inc. Washington DC, USA.
#  
#   This library is free software; you can redistribute it and/or modify it
#   under the terms of the GNU Library General Public License as published by
#   the Free Software Foundation; either version 2 of the License, or (at your
#   option) any later version.
#  
#   This library is distributed in the hope that it will be useful, but WITHOUT
#   ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
#   FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
#   License for more details.
#  
#   You should have received a copy of the GNU Library General Public License
#   along with this library; if not, write to the Free Software Foundation,
#   Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
#  
#   Correspondence concerning AIPS++ should be addressed as follows:
#          Internet email: aips2-request@nrao.edu.
#          Postal address: AIPS++ Project Office
#                          National Radio Astronomy Observatory
#                          520 Edgemont Road
#
package Path;
require 5.000;
require Exporter;
@ISA = qw(Exporter);
@EXPORT = qw(fullPath cwd syncCwd relPath Cwd CachedCwd);

sub Cwd {
    my $cwd = `pwd`;
    chop($cwd);
    $cwd =~ s|/$||;
    return $cwd;
}

sub CachedCwd {
    $cwd = Cwd() if ! $cwd;
    return $cwd;
}

sub fullPath {
    $path = shift;
    my $cwd = CachedCwd();

    $path =~ s/^\s*//;
    $path =~ s|/?\s*$||;
    if ( $path =~ /^\.\./ ) {
	$parent = $cwd;
	while ( $path =~ s|^\.\./?|| ) {
	    $parent =~ s|(.*)/[^/]+/?$|$1|;
	}
	$path = $parent . "/" . $path;
    }
    $path =~ s|^\.|$cwd|;
    $path = $cwd . "/" . $path if $path !~ m|^/|;
    return $path;
}

##
## Given two PATHS this returns the relative path from
## the first to the second.
##
sub relPath {
    my $from = fullPath(shift);
    my $to = fullPath(shift);
    my @from = split(/\//,$from);
    my @to = split(/\//,$to);

    while ( scalar(@from) && scalar(@to) ) {
	$from = shift(@from);
	$to = shift(@to);
	if ( $from ne $to ) {
	    unshift(@from,$from);
	    unshift(@to,$to);
	    last;
	}
    }

    if ( scalar(@to) && ! scalar(@from) ) {
	return join('/',@to);
    } elsif ( scalar(@to) && scalar(@from) ) {
	$from = '../' x scalar(@from);
	return $from . join('/',@to);
    } elsif ( ! scalar(@to) && scalar(@from) ) {
	$from = '../' x (scalar(@from)-1) . '..';
	return $from;
    }
    return '';
}

sub cwd {
    if (! $cwd ) {
	$cwd = `pwd`;
	chop($cwd);
	$cwd =~ s|/$||;
    }
    return $cwd;
}

sub syncCwd {
    $cwd = `pwd`;
    chop($cwd);
    $cwd =~ s|/$||;
}
1;
