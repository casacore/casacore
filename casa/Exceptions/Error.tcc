//# Error.cc: Base class for all AIPS++ errors (templated classes)
//# Copyright (C) 1993,1994,1995,1996,2000,2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <casa/Exceptions/Error.h>
#include <casa/iostream.h>


namespace casa { //# NAMESPACE CASA - BEGIN

template<class t>
indexError<t>::indexError( t oI, const Char *str, Category c )
: IndexError(str,c),
  oIndex    (oI)
{}

template<class t>
indexError<t>::indexError( t oI, const String &str, Category c )
: IndexError(str,c),
  oIndex    (oI)
{}

template<class t>
indexError<t>::~indexError() throw()
{}


template<class t>
duplError<t>::duplError(t oI, const Char *str,Category c)
: DuplError(str,c),
  oKey     (oI)
{}

template<class t>
duplError<t>::duplError(t oI, const String &str,Category c)
: DuplError(str,c),
  oKey     (oI)
{}

template<class t>
duplError<t>::~duplError() throw()
{}


} //# NAMESPACE CASA - END

