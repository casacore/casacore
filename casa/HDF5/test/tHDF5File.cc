//# tHDF5File.cc: Test program for class HDF5File
//# Copyright (C) 2008
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <casa/HDF5/HDF5File.h>
#include <casa/OS/Path.h>
#include <casa/Utilities/Assert.h>
#include <casa/Exceptions/Error.h>

using namespace casa;

int main()
{
  // Exit with untested if no HDF5 support.
  if (! HDF5Object::hasHDF5Support()) {
    return 3;
  }
  try {
    {
      // Create the file.
      HDF5File file("tHDF5File_tmpx", ByteIO::New);
      AlwaysAssertExit (Path(file.getName()).baseName() == "tHDF5File_tmpx");
      AlwaysAssertExit (! file.isClosed());
      AlwaysAssertExit (file.isWritable());
      AlwaysAssertExit (! file.isOpenedForDelete());
      // Close it temporarily.
      file.close();
      AlwaysAssertExit (file.isClosed());
      AlwaysAssertExit (file.isWritable());
      AlwaysAssertExit (! file.isOpenedForDelete());
      // Flush is a no-op if the file is closed.
      file.flush();
      // Reopen it.
      file.reopen();
      AlwaysAssertExit (! file.isClosed());
      AlwaysAssertExit (file.isWritable());
      AlwaysAssertExit (! file.isOpenedForDelete());
      file.flush();
    }
    {
      // Open readonly.
      HDF5File file("tHDF5File_tmpx", ByteIO::Old);
      AlwaysAssertExit (! file.isClosed());
      AlwaysAssertExit (! file.isWritable());
      AlwaysAssertExit (! file.isOpenedForDelete());
      // Reopen for read/write.
      file.reopenRW();
      AlwaysAssertExit (! file.isClosed());
      AlwaysAssertExit (file.isWritable());
      AlwaysAssertExit (! file.isOpenedForDelete());
    }
    {
      // Open for delete.
      HDF5File file("tHDF5File_tmpx", ByteIO::Delete);
      AlwaysAssertExit (! file.isClosed());
      AlwaysAssertExit (file.isWritable());
      AlwaysAssertExit (file.isOpenedForDelete());
    }
    {
      // File name gets absolute, so ignore for output comparison.
      cout << ">>>" << endl;
      Bool succ = True;
      try {
	HDF5File file("tHDF5File_tmpx", ByteIO::Old);
      } catch (AipsError& x) {
	succ = False;
	cout << x.what() << endl;
      }
      AlwaysAssertExit (!succ);
      cout << "<<<" << endl;
    }
    {
      // Create the file for scratch.
      HDF5File file("tHDF5File_tmp", ByteIO::Scratch);
      AlwaysAssertExit (! file.isClosed());
      AlwaysAssertExit (file.isWritable());
      AlwaysAssertExit (file.isOpenedForDelete());
    }
    {
      // Create the file. Succeeds, because scratch file is deleted.
      HDF5File file("tHDF5File_tmp", ByteIO::NewNoReplace);
    }
    {
      // Create the file. Fails, because already exists.
      cout << ">>>" << endl;
      Bool succ = True;
      try {
      HDF5File file("tHDF5File_tmp", ByteIO::NewNoReplace);
      } catch (AipsError& x) {
	succ = False;
	cout << x.what() << endl;
      }
      AlwaysAssertExit (!succ);
      cout << "<<<" << endl;
    }

  } catch (AipsError& x) {
    cout << "Unexpected exception: " << x.what() << endl;
    return 1;
  }
  cout << "OK" << endl;
  return 0;
}
