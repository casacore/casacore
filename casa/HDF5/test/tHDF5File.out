>>>
HDF5 error: Could not open or create HDF5 file tHDF5File_tmpx
<<<
>>>
HDF5 error: Could not open or create HDF5 file tHDF5File_tmp
<<<
OK
